core/src/mast/multi_forest_node_iterator.rs

```diff
@@ -0,0 +1,395 @@
+            *node_visited_mut = true;
+        }
+    }
+
+    /// Discovers a tree starting at the given root index.
+    fn discover_tree(
+        &mut self,
+        forest_idx: usize,
+        root_idx: MastNodeId,
+    ) -> Result<(), MastForestError> {
+        let current_node =
+            &self.mast_forests[forest_idx].nodes.get(root_idx.as_usize()).ok_or_else(|| {
+                MastForestError::NodeIdOverflow(
+                    root_idx,
+                    self.mast_forests[forest_idx].num_nodes() as usize,
+                )
+            })?;
+
+        // Note that the order in which we add or discover nodes is the reverse of postorder, since
+        // we're pushing them onto a stack, which reverses the order itself. Hence, reversing twice
+        // gives us the actual postorder we want.
+        match current_node {
+            MastNode::Block(_) => {
+                self.mark_for_visit(forest_idx, root_idx);
+            },
+            MastNode::Join(join_node) => {
+                self.mark_for_visit(forest_idx, root_idx);
+                self.discover_tree(forest_idx, join_node.second())?;
+                self.discover_tree(forest_idx, join_node.first())?;
+            },
+            MastNode::Split(split_node) => {
+                self.mark_for_visit(forest_idx, root_idx);
+                self.discover_tree(forest_idx, split_node.on_false())?;
+                self.discover_tree(forest_idx, split_node.on_true())?;
+            },
+            MastNode::Loop(loop_node) => {
+                self.mark_for_visit(forest_idx, root_idx);
+                self.discover_tree(forest_idx, loop_node.body())?;
+            },
+            MastNode::Call(call_node) => {
+                self.mark_for_visit(forest_idx, root_idx);
+                self.discover_tree(forest_idx, call_node.callee())?;
+            },
+            MastNode::Dyn(_) => {
+                self.mark_for_visit(forest_idx, root_idx);
+            },
+            MastNode::External(external_node) => {
+                if let Some((other_forest_idx, other_node_id)) =
+                    self.non_external_nodes.get(&external_node.digest()).copied()
+                {
+                    let visited = self.node_visited[forest_idx]
+                        .get(root_idx.as_usize())
+                        .expect("node_visited can be safely indexed by any valid MastNodeId");
+                    if !visited {
+                        self.unvisited_node_stack.push(
+                            MultiMastForestIteratorItem::ExternalNodeReplacement {
+                                replacement_forest_idx: other_forest_idx,
+                                replacement_mast_node_id: other_node_id,
+                                replaced_forest_idx: forest_idx,
+                                replaced_mast_node_id: root_idx,
+                            },
+                        );
+                    }
```

Comment:
Could we add some comments here? My understanding is that when visiting an external node, if this node has a non-external counterpart, we build this `ExternalNodeReplacement` variant which contains info of both nodes - right?

Reply:
I have addressed all comments now and documented the code. Let me know if anything is not understandable and needs a better explanation.